Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,11 +4,113 @@
 #include "Widgets/PSOverlayWidget.h"
 
 #include "Components/Image.h"
+#include "Curves/CurveFloat.h"
+#include "Data/PSDataAsset.h"
 
 #include UE_INLINE_GENERATED_CPP_BY_NAME(PSOverlayWidget)
 
-// Sets the visibility of the background overlay and lock icon.
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		ESlateVisibility PrevOverlayOpacity = PSCBackgroundOverlay->GetVisibility();
+		ESlateVisibility PrevIconOpacity = PSCBackgroundIconLock->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate && PrevIconOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+	if (bShouldPlayFadeAnimationInternal)
+	{
+		const bool bIsFinished = !FadeOverlayElementsAnimation(StartTimeFadeAnimationInternal);
+		if (bIsFinished)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+			StartTimeFadeAnimationInternal = 0.f;
+			if (!bIsFadeInAnimationInternal)
+			{
+				SetOverlayItemsVisibility(ESlateVisibility::Collapsed);
+			}
+		}
+	}
+}
+
+// Event to execute when widget is ready
+void UPSOverlayWidget::NativeConstruct()
+{
+	Super::NativeConstruct();
+	FadeCurveFloatInternal = UPSDataAsset::Get().GetFadeCurveFloat();
```

Comment:
why do you need to cache it? `UPSDataAsset::Get().GetFadeCurveFloat();` is simple getter that has no overhead. It does not search anything, but obtains curve directly from data asset. Please use `UPSDataAsset::Get().GetFadeCurveFloat();` everywhere and remove redundant property to have less properties to maintain

Reply:
the idea was only to reduce that call. Removed cache as part of: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/5ed0d922beeddc33284fb15e6ceb3cb284603623